File: packages/babel-plugin/src/utils/is-compiled.ts
import * as t from '@babel/types';

import type { State } from '../types';

/**
 * Returns `true` if the node is using `css` from `@compiled/react` as a call expression
 *
 * @param node {t.Expression} The expression node that is being checked
 * @param state {State} Plugin state
 * @returns {boolean} Whether the node is a css usage from compiled
 */
export const isCompiledCSSCallExpression = (
  node: t.Expression,
  state: State
): node is t.CallExpression =>
  t.isCallExpression(node) &&
  t.isIdentifier(node.callee) &&
  [...(state.compiledImports?.css || []), state.importedCompiledImports?.css].includes(
    node.callee.name
  );

/**
 * Returns `true` if the node is using `css` from `@compiled/react` as a tagged template expression
 *
 * @param node {t.Expression} The expression node that is being checked
 * @param state {State} Plugin state
 * @returns {boolean} Whether the node is a css usage from compiled
 */
export const isCompiledCSSTaggedTemplateExpression = (
  node: t.Expression,
  state: State
): node is t.TaggedTemplateExpression =>
  t.isTaggedTemplateExpression(node) &&
  t.isIdentifier(node.tag) &&
  !!state.compiledImports?.css?.includes(node.tag.name);

/**
 * Returns `true` if the node is using `keyframes` from `@compiled/react` as a call expression
 *
 * @param node {t.Node} The node that is being checked
 * @param state {State} Plugin state
 * @returns {boolean} Whether the node is a compiled keyframe
 */
export const isCompiledKeyframesCallExpression = (
  node: t.Node,
  state: State
): node is t.CallExpression =>
  t.isCallExpression(node) &&
  t.isIdentifier(node.callee) &&
  !!state.compiledImports?.keyframes?.includes(node.callee.name);

/**
 * Returns `true` if the node is using `cssMap` from `@compiled/react` as a call expression
 *
 * @param node {t.Node} The node that is being checked
 * @param state {State} Plugin state
 * @returns {boolean} Whether the node is a compiled cssMap
 */
export const isCompiledCSSMapCallExpression = (
  node: t.Node,
  state: State
): node is t.CallExpression =>
  t.isCallExpression(node) &&
  t.isIdentifier(node.callee) &&
  !!state.compiledImports?.cssMap?.includes(node.callee.name);

/**
 * Returns `true` if the node is using `keyframes` from `@compiled/react` as a tagged template expression
 *
 * @param node {t.Node} The node that is being checked
 * @param state {State} Plugin state
 * @returns {boolean} Whether the node is a compiled keyframe
 */
export const isCompiledKeyframesTaggedTemplateExpression = (
  node: t.Node,
  state: State
): node is t.TaggedTemplateExpression =>
  t.isTaggedTemplateExpression(node) &&
  t.isIdentifier(node.tag) &&
  !!state.compiledImports?.keyframes?.includes(node.tag.name);

/**
 * Returns `true` if the node is using `styled` from `@compiled/react` from a styled.tag usage
 *
 * @param node {t.Node} The node that is being checked
 * @param state {State} Plugin state
 * @returns {boolean} Whether the node is a styled usage from compiled
 */
const isCompiledStyledMemberExpression = (node: t.Node, state: State): node is t.MemberExpression =>
  t.isMemberExpression(node) &&
  t.isIdentifier(node.object) &&
  !!state.compiledImports?.styled?.includes(node.object.name);

/**
 * Returns `true` if the node is using `styled` from `@compiled/react` from a styled(Component) usage
 *
 * @param node {t.Node} The node that is being checked
 * @param state {State} Plugin state
 * @returns {boolean} Whether the node is a styled usage from compiled
 */
const isCompiledStyledCompositionCallExpression = (
  node: t.Node,
  state: State
): node is t.CallExpression =>
  t.isCallExpression(node) &&
  t.isIdentifier(node.callee) &&
  !!state.compiledImports?.styled?.includes(node.callee.name);

/**
 * Returns `true` if the node is using `styled` from `@compiled/react` as a call expression
 *
 * @param node {t.Expression} The expression node that is being checked
 * @param state {State} Plugin state
 * @returns {boolean} Whether the node is a styled usage from compiled
 */
export const isCompiledStyledCallExpression = (
  node: t.Expression,
  state: State
): node is t.CallExpression =>
  t.isCallExpression(node) &&
  (isCompiledStyledMemberExpression(node.callee, state) ||
    isCompiledStyledCompositionCallExpression(node.callee, state));

/**
 * Returns `true` if the node is using `styled` from `@compiled/react` as a tagged template expression
 *
 * @param node {t.Node} The node that is being checked
 * @param state {State} Plugin state
 * @returns {boolean} Whether the node is a styled usage from compiled
 */
export const isCompiledStyledTaggedTemplateExpression = (
  node: t.Node,
  state: State
): node is t.TaggedTemplateExpression =>
  t.isTaggedTemplateExpression(node) &&
  (isCompiledStyledMemberExpression(node.tag, state) ||
    isCompiledStyledCompositionCallExpression(node.tag, state));

/** Returns true if and only if the current node is one of the following:
 * - A usage of the `css` API from Compiled
 * - A usage
 * - A usage of the `styled` API from Compiled
 * - IS NOT a usage of the `keyframes` API from Compiled
 */
export const isCompiledStyleCall = (node: t.Expression, state: State): boolean => {
  const checkers = [
    isCompiledCSSCallExpression,
    isCompiledCSSTaggedTemplateExpression,
    isCompiledCSSMapCallExpression,
    isCompiledStyledCompositionCallExpression,
    isCompiledStyledCallExpression,
    isCompiledStyledTaggedTemplateExpression,
  ];

  return checkers.some((checker) => checker(node, state));
};

File: packages/swc-plugin/src/utils/is_compiled.rs
use swc_core::ecma::ast::*;
use crate::types::TransformState;

/// Check if a call expression is a compiled CSS call expression
pub fn is_compiled_css_call_expression(expr: &Expr, state: &TransformState) -> bool {
    if let Expr::Call(call) = expr {
        return is_compiled_call(call, "css", state);
    }
    false
}

/// Check if a tagged template expression is a compiled CSS tagged template
pub fn is_compiled_css_tagged_template_expression(expr: &Expr, state: &TransformState) -> bool {
    if let Expr::TaggedTpl(tagged) = expr {
        return is_compiled_tagged_template(tagged, "css", state);
    }
    false
}

/// Check if a call expression is a compiled styled call expression
pub fn is_compiled_styled_call_expression(expr: &Expr, state: &TransformState) -> bool {
    if let Expr::Call(call) = expr {
        return is_compiled_call(call, "styled", state);
    }
    false
}

/// Check if a tagged template expression is a compiled styled tagged template
pub fn is_compiled_styled_tagged_template_expression(expr: &Expr, state: &TransformState) -> bool {
    if let Expr::TaggedTpl(tagged) = expr {
        return is_compiled_tagged_template(tagged, "styled", state);
    }
    false
}

/// Check if a call expression is a compiled keyframes call expression
pub fn is_compiled_keyframes_call_expression(expr: &Expr, state: &TransformState) -> bool {
    if let Expr::Call(call) = expr {
        return is_compiled_call(call, "keyframes", state);
    }
    false
}

/// Check if a tagged template expression is a compiled keyframes tagged template
pub fn is_compiled_keyframes_tagged_template_expression(expr: &Expr, state: &TransformState) -> bool {
    if let Expr::TaggedTpl(tagged) = expr {
        return is_compiled_tagged_template(tagged, "keyframes", state);
    }
    false
}

/// Check if a call expression is a compiled CSS map call expression
pub fn is_compiled_css_map_call_expression(expr: &Expr, state: &TransformState) -> bool {
    if let Expr::Call(call) = expr {
        return is_compiled_call(call, "cssMap", state);
    }
    false
}

/// Check if a call expression is a compiled API call
fn is_compiled_call(call: &CallExpr, api_name: &str, state: &TransformState) -> bool {
    match &call.callee {
        Callee::Expr(expr) => {
            match expr.as_ref() {
                Expr::Ident(ident) => {
                    // Check if this identifier is imported from compiled
                    if let Some(compiled_imports) = &state.compiled_imports {
                        match api_name {
                            "css" => compiled_imports.css.as_ref().map_or(false, |imports| imports.contains(&ident.sym.to_string())),
                            "styled" => compiled_imports.styled.as_ref().map_or(false, |imports| imports.contains(&ident.sym.to_string())),
                            "keyframes" => compiled_imports.keyframes.as_ref().map_or(false, |imports| imports.contains(&ident.sym.to_string())),
                            "cssMap" => compiled_imports.css_map.as_ref().map_or(false, |imports| imports.contains(&ident.sym.to_string())),
                            _ => false,
                        }
                    } else {
                        // If no compiled imports tracked, check default name
                        ident.sym.as_ref() == api_name
                    }
                }
                Expr::Member(member) => {
                    // Handle styled.div, styled.span etc.
                    if api_name == "styled" {
                        if let Expr::Ident(obj) = member.obj.as_ref() {
                            if let Some(compiled_imports) = &state.compiled_imports {
                                compiled_imports.styled.as_ref().map_or(false, |imports| imports.contains(&obj.sym.to_string()))
                            } else {
                                obj.sym.as_ref() == "styled"
                            }
                        } else {
                            false
                        }
                    } else {
                        false
                    }
                }
                _ => false,
            }
        }
        _ => false,
    }
}

/// Check if a tagged template is a compiled API tagged template
fn is_compiled_tagged_template(tagged: &TaggedTpl, api_name: &str, state: &TransformState) -> bool {
    match tagged.tag.as_ref() {
        Expr::Ident(ident) => {
            // Check if this identifier is imported from compiled
            if let Some(compiled_imports) = &state.compiled_imports {
                match api_name {
                    "css" => compiled_imports.css.as_ref().map_or(false, |imports| imports.contains(&ident.sym.to_string())),
                    "styled" => compiled_imports.styled.as_ref().map_or(false, |imports| imports.contains(&ident.sym.to_string())),
                    "keyframes" => compiled_imports.keyframes.as_ref().map_or(false, |imports| imports.contains(&ident.sym.to_string())),
                    _ => false,
                }
            } else {
                // If no compiled imports tracked, check default name
                ident.sym.as_ref() == api_name
            }
        }
        Expr::Member(member) => {
            // Handle styled.div``, styled.span`` etc.
            if api_name == "styled" {
                if let Expr::Ident(obj) = member.obj.as_ref() {
                    if let Some(compiled_imports) = &state.compiled_imports {
                        compiled_imports.styled.as_ref().map_or(false, |imports| imports.contains(&obj.sym.to_string()))
                    } else {
                        obj.sym.as_ref() == "styled"
                    }
                } else {
                    false
                }
            } else {
                false
            }
        }
        _ => false,
    }
}

/// Check if a JSX function call is transformed (for error detection)
pub fn is_transformed_jsx_function(expr: &Expr, _state: &TransformState) -> bool {
    if let Expr::Call(call) = expr {
        if let Callee::Expr(callee) = &call.callee {
            if let Expr::Ident(ident) = callee.as_ref() {
                return ident.sym.as_ref() == "jsx" || ident.sym.as_ref() == "_jsx";
            }
        }
    }
    false
}